pre_commit/languages/download.py

```diff
@@ -0,0 +1,258 @@
+from __future__ import annotations
+
+import contextlib
+import hashlib
+import os.path
+import platform
+import stat
+from base64 import standard_b64decode as b64decode
+from base64 import standard_b64encode as b64encode
+from dataclasses import dataclass
+from os import chmod
+from pathlib import Path
+from pathlib import PurePath
+from typing import AnyStr
+from typing import Generator
+from typing import IO
+from typing import Iterator
+from typing import Literal
+from typing import Sequence
+from urllib.parse import urlparse
+from urllib.request import urlopen
+
+from pre_commit import lang_base
+from pre_commit.envcontext import envcontext
+from pre_commit.envcontext import PatchesT
+from pre_commit.envcontext import Var
+from pre_commit.prefix import Prefix
+
+ENVIRONMENT_DIR = 'download'
+get_default_version = lang_base.basic_get_default_version
+run_hook = lang_base.basic_run_hook
+
+
+def get_env_patch(target_dir: str) -> PatchesT:
+    return (
+        ('PATH', (target_dir, os.pathsep, Var('PATH'))),
+    )
+
+
+@contextlib.contextmanager
+def in_env(prefix: Prefix, version: str) -> Generator[None, None, None]:
+    envdir = lang_base.environment_dir(prefix, ENVIRONMENT_DIR, version)
+    with envcontext(get_env_patch(envdir)):
+        yield
+
+
+@dataclass(frozen=True)
+class Platform:
+    value: str
+
+    def __post_init__(self) -> None:
+        os, cpu = self.parts
+
+        if os not in {'linux', 'darwin', 'windows'}:
+            raise ValueError(f'invalid operating system `{os}`')
+
+        if cpu not in {'amd64', 'aarch64'}:
+            raise ValueError(f'invalid CPU `{cpu}`')
+
+    @property
+    def parts(self) -> tuple[str, str]:
+        first, second = self.value.split('/', 1)
+        return (first, second)
+
+    @property
+    def os(self) -> Literal[
+        'linux', 'darwin', 'windows', 'dragonfly',
+        'freebsd',
+    ]:
+        os, _ = self.parts
+        return os  # type: ignore
+
+    @property
+    def cpu(self) -> Literal[
+        'arm64', 'arm64be', 'arm', '386', 'amd64', 'ppc',
+        'ppc64', 'ppc64le',
+    ]:
+        _, cpu = self.parts
+        return cpu  # type: ignore
+
+    @staticmethod
+    def host() -> Platform:
+        os = {
+            'Linux': 'linux',
+            'Darwin': 'darwin',
+            'Windows': 'windows',
+            'DragonFly': 'dragonfly',
+            'FreeBSD': 'freebsd',
+        }[platform.system()]
+        cpu = {
+            'aarch64': 'arm64',
+            'aarch64_be': 'arm64be',
+            'arm': 'arm',
+            'i386': '386',
+            'i686': '386',
+            'x86_64': 'amd64',
+            'ppc': 'ppc',
+            'ppc64': 'ppc64',
+            'ppc64le': 'ppc64le',
+        }[platform.machine()]
+        return Platform(f'{os}/{cpu}')
+
+
+class ContentCorruptionError(Exception):
+    pass
```

Comment:
* It is often useful to have the objects that raised the exception attached to the error to allow customised handling.
* It is not guaranteed that corruption has occurred, the purpose of this error is to signify a incorrect checksum.
* It is best to inherit from a common built-in exception if possible, `ValueError` seems sensible here.

```suggestion
from os import linesep

class Hash(Protocol):
    @property
    def digest_size(self) -> int: ...
    @property
    def block_size(self) -> int: ...
    @property
    def name(self) -> str: ...
    def digest(self) -> bytes: ...
    def hexdigest(self) -> str: ...

class ChecksumError(ValueError):
    def __init__(self, sri: SRI, hash: Hash, message: str = "Mismatched checksum"):
        self.__sri = sri
        self.__hash = hash
        self.__message = message

    @property
    def sri(self) -> SRI:
        return self.__sri

    @property
    def hash(self) -> Hash:
        return self.__hash

    @property
    def message(self) -> str:
        return self.__message

    def __str__(self) -> str:
        return f"{self.__message}:{linesep} - expected: {self.sri.hexdigest(){linesep} - actual  : {self.hash.hexdigest()
```

Reply:
This is my understanding of the terminology - I think ``digest`` is a property of any byte array.  ``checksum`` is the signature of a particular byte array - if that byte array is correct, the array's ``digest`` should be the same as ``checksum``. ``hashlib`` can output a ``digest`` at any time, but we check if the ``digest`` is the same as the ``checksum``. 

I think it is not really useful to expose ``hexdigest``, the checksum in additional dependencies is encoded in base64. For comparison purposes, it is sufficient to compare against the encoded base64 checksum rather than the raw byte array. In the new design, the hasher within the SRI object is now ephemeral - it is a temporary variable in the ``check`` call, so I don't think it makes sense to have a ``digest`` property. 

Let me know what you think.